pandas/core/frame.py

```diff
@@ -440,6 +443,8 @@ def __init__(self, data=None, index=None, columns=None, dtype=None, copy=False):
             if not isinstance(data, abc.Sequence):
                 data = list(data)
             if len(data) > 0:
+                if is_dataclass_instance(data[0]):
```

Comment:
We need to document this behavior (that the first item is inspected for a dataclass) somewhere. Probably in `dsintro.rst`.

Reply:
It's the same as with dicts, or Series, ... (so would need to check if that is documented, otherwise need to add documentation in general about this).

Reply:
where is the proper place to add this to the documentation? this is my first contribution

Reply:
These cases seem to be documented in `getting_started/dsintro.rst`. A short section on "From a list of Dataclasses" would be nice.

Reply:
@TomAugspurger I just noticed, there is no entry in that file for NamedTuples. The conversion for namedtuples is similar to what I am doing, the constructor checks the first entry and uses the fields provided by the element as a generalization for the rest. I will update the documentation for the dataclasses sometime tomorrow, but do you think we should open a PR for the namedtuples?

Reply:
> and uses the fields provided by the element as a generalization for the rest.

This is something that is actually buggy IMO, and so best not to mention it in the docs. See https://github.com/pandas-dev/pandas/issues/27329 (but a PR to fix this is certainly welcome)

Reply:
I will try and see, I think that could be solved by converting all namedtuples to dict as well, that way they could leach off the implementation as well.